### DISCUSSION TITLE
State estimations using Pybamm models

### DISCUSSION THREAD

**Discussion by ssingh-ipa** [pybamm-team/PyBaMM#1486] [Q&A]
Hello,

I am interested in estimating the state parameters of a battery using the DFN model, i.e. State of charge, state of health, state of available power, state of function, etc. Basically any of the parameters that are needed to estimate the remaining useful life of a battery. I am guessing such discussions might already exist in this forum and hopefully I can be directed to those discussions.  

Here is a brief reference to what I am trying to do using pybamm - I would provide the experimental procedure (or test procedures generally done for characterization of batteries in a cell tester) to the model. Lets say, by using the CC method or OCV method, we calculate the SOC of the battery using the experimental data. Next we would provide driving cycle data to the model and again calculate the SOC and see the change in its value due to that particular driving cycle. 

I am open to suggestions and feedback from the developers team. Is this even possible or are there any alternatives?

**Comment by valentinsulzer** [comment]
There are two parts to this:
1. Define SOC (and SOH, SOF, etc) as states in the model. SOC will be pretty close after #1459 gets merged. SOH is trickier since it doesn't have a single agreed-upon definition, but after #1459 you'll be able to track capacity, internal resistance, LAM, LLI, etc per cycle.
2. State estimation using e.g. filtering. To be honest this isn't something we've looked at with pybamm. It would be a big undertaking but definitely interesting. We already use [casadi](https://web.casadi.org/) for solving the ODEs/DAEs, so that would be the way to go

**Reply by ssingh-ipa** [reply]
So, is the experiment-refractor is not a part of the package yet? Rather I can expect it to be in integrated in the next few months.
I have been working with battery modelling through MATLAB for state estimations, however within the scope of a project for intelligent BMS and digital twins, we are exploring other methods for applied research. I see a good potential of collaboration with pybamm. Please let me know if I can get in touch with someone from pybamm in order to discuss this further.

**Reply by valentinsulzer** [reply]
It is almost ready to be merged into the develop branch, just needs an example to show how to use the new functionality.
Then you can use it if you install from source, but have to wait until the next release (in a few months) to pip install it.
You can join the slack here to talk about state estimation https://join.slack.com/t/pybamm/shared_invite/zt-pq0k750g-pOrH1_i_qB_AlLLlWBB2Eg

**Reply by Mrzhang-hub** [reply]
> It is almost ready to be merged into the develop branch, just needs an example to show how to use the new functionality. Then you can use it if you install from source, but have to wait until the next release (in a few months) to pip install it. You can join the slack here to talk about state estimation https://join.slack.com/t/pybamm/shared_invite/zt-pq0k750g-pOrH1_i_qB_AlLLlWBB2Eg

Sir, can you invite me to the stack community to discuss the topic of state estimation through pybamm? I am studying how to design EKF to estimate the internal state via pybamm.

**Reply by muchajan** [reply]
@valentinsulzer Could I also please get an invitation to the Slack channel? I'm also trying to create state space equations for SoC estimation using EKF.
